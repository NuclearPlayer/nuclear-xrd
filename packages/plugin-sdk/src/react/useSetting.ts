import { useEffect, useMemo, useRef, useState } from 'react';

import type { SettingsHost, SettingValue } from '../types/settings';

export const useSetting = <T extends SettingValue = SettingValue>(
  host: SettingsHost | undefined,
  id: string,
) => {
  const hostRef = useRef(host);
  hostRef.current = host;

  const [currentValue, setCurrentValue] = useState<T | undefined>(undefined);

  useEffect(() => {
    if (!hostRef.current) return;
    let unsubscribe: (() => void) | undefined;
    let isMounted = true;

    hostRef.current.get<T>(id).then((initialValue) => {
      if (!isMounted) return;
      setCurrentValue(initialValue);
      unsubscribe = hostRef.current?.subscribe<T>(id, (nextValue) =>
        setCurrentValue(nextValue),
      );
    });

    return () => {
      isMounted = false;
      if (unsubscribe) unsubscribe();
    };
  }, [id, hostRef]);

  const setValue = useMemo(
    () => (nextValue: T) => {
      if (!hostRef.current) return;
      void hostRef.current.set<T>(id, nextValue);
    },
    [id],
  );

  return [currentValue, setValue] as const;
};
