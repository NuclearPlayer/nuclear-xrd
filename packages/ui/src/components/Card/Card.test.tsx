import { render } from '@testing-library/react';

import { Card } from '.';

describe('Card', () => {
  it('(Snapshot) renders correctly', async () => {
    const { container } = render(
      <Card
        src="https://picsum.photos/300"
        title="Test album"
        subtitle="Test artist"
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
