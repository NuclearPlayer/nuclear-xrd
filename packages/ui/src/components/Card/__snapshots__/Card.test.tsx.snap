// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card > (Snapshot) renders correctly 1`] = `
<div>
  <button
    class="cursor-pointer rounded-md transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none h-auto flex w-42 flex-col items-stretch gap-2 p-2 text-left"
    data-headlessui-state=""
    data-testid="card"
    type="submit"
  >
    <div
      class="border-border text-foreground flex h-full rounded-md border-2 bg-primary shadow-none relative aspect-square w-full items-center justify-center overflow-hidden p-0"
    >
      <div
        class="overflow-hidden absolute inset-0"
      >
        <img
          alt="Test album"
          class="h-full w-full object-cover"
          decoding="async"
          loading="lazy"
          src="https://picsum.photos/300"
        />
        <div
          class="pointer-events-none absolute inset-0 flex items-center justify-center"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-cassette-tape animate-pulse opacity-70"
            fill="none"
            height="96"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="0.5"
            viewBox="0 0 24 24"
            width="96"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="16"
              rx="2"
              width="20"
              x="2"
              y="4"
            />
            <circle
              cx="8"
              cy="10"
              r="2"
            />
            <path
              d="M8 12h8"
            />
            <circle
              cx="16"
              cy="10"
              r="2"
            />
            <path
              d="m6 20 .7-2.9A1.4 1.4 0 0 1 8.1 16h7.8a1.4 1.4 0 0 1 1.4 1l.7 3"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="min-w-0"
    >
      <div
        class="text-foreground truncate text-sm font-bold"
      >
        Test album
      </div>
      <div
        class="text-foreground-secondary truncate text-xs"
      >
        Test artist
      </div>
    </div>
  </button>
</div>
`;
